src/samVisualize/rendering/forceDirectedGraph.ts

```diff
@@ -0,0 +1,703 @@
+/*!
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+import * as RenderConstants from '../rendering/renderConstants'
+import { RenderedLinkTypes } from '../samVisualizeTypes'
+import * as d3 from 'd3'
+import { WebviewApi } from 'vscode-webview'
+import { GraphObject } from '../graphGeneration/graph'
+import { filterPrimaryOnly } from '../rendering/filter'
+import * as _ from 'lodash'
+
+interface NodeDatum extends d3.SimulationNodeDatum {
+    name: string
+    type?: string
+}
+
+interface LinkDatum extends d3.SimulationLinkDatum<NodeDatum> {
+    source: string | NodeDatum
+    target: string | NodeDatum
+    type?: string
+}
+
+export class ForceDirectedGraph {
+    /**
+     * The complete, unfiltered GraphObject
+     */
+    private completeGraphObject: GraphObject | undefined
+
+    /**
+     * A subset of the completeGraphObject, containing only primary nodes and links between them.
+     */
+    private primaryOnlyGraphObject: GraphObject | undefined
+
+    /**
+     * Maps a resource type identifier to a WebviewURI string containing the corresponding icon
+     */
+    private iconPaths: { [resourceType: string]: string }
+
+    private vscodeApi: WebviewApi<unknown> | undefined
+
+    // The overarching g container containing nodes & links
+    // Having this separate container element allows for zooms and pans
+    private g: d3.Selection<SVGGElement, unknown, HTMLElement | null, unknown>
+
+    // The d3.Simulation which runs the physics and interactivity of the force directed graph
+    public simulation: d3.Simulation<NodeDatum, LinkDatum>
+
+    // Since the link construction uses the id of the arrowheads,
+    // it's declared and passed around to avoid mismatching string literals.
+    public readonly arrowheadID = 'arrowhead'
+
+    private links: d3.Selection<SVGGElement, LinkDatum, SVGGElement, unknown> | undefined
+    private nodes: d3.Selection<SVGGElement, NodeDatum, SVGGElement, unknown> | undefined
+
+    // Holds the div containing the filter radio buttons
+    // Visibility is toggled during error renders
+    private filterButtonsDiv: d3.Selection<HTMLDivElement, unknown, HTMLElement | null, unknown>
+
+    public constructor(
+        completeGraphObject: GraphObject,
+        primaryResourceList: string[],
```

Comment:
Can it be a `Set` from the start?

Reply:
The thing is, VSCode webviews load their HTML using a string. Which is not ideal. This means that whatever parameters I want to pass into any functions called within my webview, they need to be string literals. Since JS doesn't have a Set literal, I had to use a list literal to get that param into the webview. There may be other ways to do it, but I could not find any obvious alternatives. I can look deeper into it, but I have not seen any webview examples that were not loading a big HTML string.